storage-s3-dynamodb/src/test/scala/io/delta/storage/ExternalLogStoreSuite.scala

```diff
@@ -57,61 +65,92 @@ class ExternalLogStoreSuite extends org.apache.spark.sql.delta.PublicLogStoreSui
   }
 
   test("overwrite") {
-    withTempDir { tempDir =>
+    withTempLogDir { tempLogDir =>
       val store = createLogStore(spark)
-      val path = FileNames.deltaFile(new Path(new Path(tempDir.toURI), "_delta_log"), 0)
+      val path = getDeltaVersionPath(tempLogDir, 0)
       store.write(path, Iterator("foo", "bar"), overwrite = false, sessionHadoopConf)
       assert(MemoryLogStore.containsKey(path))
       store.write(path, Iterator("foo", "bar"), overwrite = true, sessionHadoopConf)
       assert(MemoryLogStore.containsKey(path))
     }
   }
 
-  test("recovery") {
+  test("write N+1 recovers version N") {
     withSQLConf(
       "fs.failing.impl" -> classOf[FailingFileSystem].getName,
       "fs.failing.impl.disable.cache" -> "true"
     ) {
-      withTempDir { tempDir =>
+      withTempLogDir { tempLogDir =>
         val store = createLogStore(spark)
-        val path = FileNames.deltaFile(
-          new Path(new Path(s"failing:${tempDir.getCanonicalPath}"), "_delta_log"),
-          0
-        )
+
+        val delta0_normal = getDeltaVersionPath(tempLogDir, 0)
+        val delta0_fail = getFailingDeltaVersionPath(tempLogDir, 0)
+        val delta1 = getDeltaVersionPath(tempLogDir, 1)
+
+        // Case 1: write N+1 fails
+        // - entry for delta0 doesn't exist in external store or in file system
+        val e = intercept[java.nio.file.FileSystemException] {
+          store.write(delta1, Iterator("one"), overwrite = false, sessionHadoopConf)
+        }
+        assert(e.getMessage == s"previous commit $delta0_normal doesn't exist")
+
+        // Case 2: write N+1 succeeds
+        // - entry for delta0 does exist in external store but not in file system
+        FailingFileSystem.failOnSuffix = Some(delta0_fail.getName)
+        store.write(delta0_fail, Iterator("zero"), overwrite = false, sessionHadoopConf)
+        assert(!delta0_fail.getFileSystem(sessionHadoopConf).exists(delta0_fail))
+        assert(!MemoryLogStore.get(delta0_fail).complete)
+        assert(MemoryLogStore.get(delta0_fail) == MemoryLogStore.get(delta0_normal))
```

Comment:
Also can you add a comment what this is checking? To me it seems it's testing the above code added to `MemoryLogStore` is correct but this might not be obvious